쿠폰, 예약 관련 엔티티 수정

matrixpower1004 opened the pull request:
### 💡Motivation
PRD에 맞추어 예약과 쿠폰 관련 엔티티 수정

### 📌Changes
- 쿠폰 관련 엔티티 신규 생성 및 수정
- 예약 관련 엔티티 수정
- QueryDSL DI 문제 해결용 configuration 추가
- 예약 상태 enum 추가

### 🫱🏻‍🫲🏻To Reviewers
QueryDSL 사용 시 Parameter 0 of constructor in xxx 에러를 해결하기 위해 configuration 파일을 추가.

Closes #32 
Closes #33

matrixpower1004 committed:
#32 feat: 예약 기능 엔티티 수정

- Reservation Entity 수정. 예약과 객실 재고를 1:1 연관 관계를 맺어주고 예약 테이블에 FK를 위치 시킴
- Payment Entity 수정. Boolean 타입의 쿠폰 사용 여부 컬럼을 생성.

matrixpower1004 committed:
#33 feat: 쿠폰 관련 엔티티 수정

- PRD에 맞추어 쿠폰 관련 엔티티 생성 및 수정
- 쿠폰 발급: CouponIssuance 엔티티 생성
- 쿠폰 사용: CouponRedeem 엔티티 생성
- Coupon 엔티티 수정

matrixpower1004 committed:
add: QueryDSL configuration

- Parameter 0 of constructor in xxx 에러 해결용 configuration

matrixpower1004 committed:
#49 refactor: PRD 변경에 따른 쿠폰 관련 엔티티 수정

- ERD 수정 반영
- 쿠폰 관련 엔티티 수정
- DiscountType enum에 view에 응답할 쿠폰 이름 만드는 로직, 할인가 반영된 결제 금액 계산하여 반환하는 로직, 할인가 검증 로직(김현아님 코드 재활용)을 응집시킴.

matrixpower1004 committed:
#49 feat: 쿠폰 & 예약 관련 repository 작성

- 쿠폰, 예약 관련 repository 작성

JeongUijeong commented on src/main/java/com/backoffice/upjuyanolja/domain/coupon/entity/Coupon.java:
```diff
@@ -33,65 +25,55 @@ public class Coupon extends BaseTime {
 
     @Column(nullable = false)
-    @Comment("유효 기간 만료일")
+    @Comment("구폰 노출 만료일")
```
오타 수정해주세요~~!
```suggestion
    @Comment("쿠폰 노출 만료일")
```

JeongUijeong commented on src/main/java/com/backoffice/upjuyanolja/domain/coupon/entity/DiscountType.java:
```diff
@@ -0,0 +1,112 @@
+    ),
+
+    RATE(
+        "%할인 쿠폰",
+        "%쿠폰",
```
별건 아니지만 % 뒤에 한 칸 공백이 있으면 좀 더 보기 편할 것 같습니다 ㅎㅎ
```suggestion
        "% 할인 쿠폰",
        "% 쿠폰",
```

matrixpower1004 replied:
현재 디자인에서는 공백 없이 붙여서 출력하도록 되어 있습니다. 변경하려면 기획, FE와 협의가 필요한 부분입니다.

JeongUijeong replied:
디자인 최종안에 다음과 같이 공백이 있습니다! ㅎㅎ

![image](https://github.com/Upjuyanolja/Upjuyanolja_BE/assets/94631526/2afe33e7-c233-4036-bf09-6873b149323b)
![image](https://github.com/Upjuyanolja/Upjuyanolja_BE/assets/94631526/0d92713a-8938-4a81-bb69-dbe1a5121bf3)
![image](https://github.com/Upjuyanolja/Upjuyanolja_BE/assets/94631526/a0ddc5de-e862-468d-ad10-c70833b0a28c)

matrixpower1004 replied:
디자인이 그새 또 바뀌었군요. 띄어쓰기 수정되었습니다.

JeongUijeong reviewed (changes requested):
수고 많으셨습니다~~! 별건 아니지만 몇 가지 소소한 코멘트 남겨드렸습니다~ 반영 가능하시다면 반영 부탁 드리겠습니다! 😊

matrixpower1004 committed:
refactor: Coupon.java 오타 수정

- 오타 수정: 구폰 -> 쿠폰

matrixpower1004 committed:
chore: DiscountType.java 띄어쓰기 수정

수정된 디자인에 맞추어 띄어쓰기 수정: %쿠폰 -> % 쿠폰

chaewon12 commented on src/main/java/com/backoffice/upjuyanolja/domain/room/entity/RoomStock.java:
```diff
@@ -32,6 +33,11 @@ public class RoomStock {
     @Comment("객실 식별자")
     private Room room;
 
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(nullable = false, name = "payment_id")
+    @Comment("결제 식별자")
+    private Payment payment;
+
```
혹시 RoomStock이 결제 식별자를 가지는 이유가 뭔가요?
객실 재고에 결제 내역이 왜 필요한지 이해가 안가서요

chaewon12 replied:
구두로 설명 들었습니다 🙂
* 코드 제거 필요

chaewon12 commented:
예약 관련 엔티티 수정이 필요해서 
제가 해당 브랜치에서 추가 작업 후 PR 업데이트 & 리뷰 재요청 드리겠습니다 !
@JeongUijeong @decten @chadongmin

chaewon12 committed:
#32 fix: '예약 내역' 이 '결제 내역'을 가지도록 관계 수정

- 결제 진행 후 최종 예약 완료하는 흐름이 자연스러움.
- 관계
  - 예약 내역 확인에서 예약정보, 결제 정보를 확인함.
  - 결제 내역 확인에서는 결제 정보만 확인함.

chaewon12 committed:
#32 fix: Payment 할인 금액 필드 추가 및 쿠폰 사용 여부 필드 제거

- 객실 금액, 할인 금액, 결제 금액을 나누어 저장
- 쿠폰 사용 여부는 '예약' 에서 관리

chaewon12 committed:
#32 fix: Reservation 과 ReservationRoom 분리

- ReservationRoom에 예약한 객실 관련 정보 저장
- Reservation에는 예약자, 이용자, 객실, 결제정보 등 전체적인 예약 관련 정보를 저장
- Reservation 상태 enum 추가

chaewon12 committed:
#32 fix: RoomStock 내 필요 없는 Payment 필드 제거

- 객실 재고에 Payment는 필요 없다

chaewon12 committed:
#32 fix: CouponRedeem이 Reservation 식별자를 가지도록 수정

- Payment -> Reservation 와 연결
- 쿠폰 수량 필드는 필요 없다고 판단하여 제거

decten commented on src/main/java/com/backoffice/upjuyanolja/domain/room/entity/RoomStock.java:
```diff
@@ -50,8 +55,7 @@ public class RoomStock {
+    public RoomStock(Long id,
         Room room,
         int count,
         RoomStatus status,
```
erd에서는 room이 roomstatus를 가지는데, roomstock로 옮겨진 이유가 궁금합니다

chaewon12 replied:
이전 erd 에서는 roomstock에 있었고 
현재 보시는 erd는 pr 이후, 의정님과 회의 에서 변경된 버전입니다!

아직 재고 테이블에 대한 확정안이 나오지 않아 (멘토님께 문의 후 확정) 코드는 변경 전입니다!
재고 테이블이 확정이 된 후 room과 roomstock 코드 수정이 이루어 질 것 같습니다 ~ 
(해당 수정은 의정님이 해주신다고 하셨어요)

decten replied:
넵 답변 감사합니다😄
수고 많으셨습니다👍👍

chaewon12 committed:
#32 style: 코드 포맷 수정 및 주석 제거

JeongUijeong reviewed (approved):
수고하셨습니다! 👍🏻👍🏻
